# Description
This page contains instructions for running this snakemake workflow on your slurm HPC. The pipeline is used to process single or paired-end reads and generates a raw gene and transcript-level count matrix that can be used for differential gene expression analysis (implementation of differential gene expression and enrichment analysis steps are currently underway). Additional outputs include trimmed fastq files, `.bam` . 
Currently the pipeline is best-suited for Illumina sequencing data. However, read-processing steps can be modified by the user by providing own adapter sequences. 

## Pipeline overview
1. Raw fastq files (.gz compressed or uncompressed) will be checked for quality using `fastqc`.
2. Trimming and adapter removal will be peformed using `fastp`.
3. A user-provided reference genome or transcriptome in FASTA format will be used for generating an index and alignment. Genome indexing and alignment will be performed using `STAR` while transcriptome indexing and mapping will be performed by `salmon`. NOTE: You do not need to run both to generate `.bam` files. 
4. To generate gene-level counts `featureCounts` from the `subread` package will be used by providing `.bam` generated by either STAR or Salmon and a genome annotation file in the GTF format as inputs.
5. To generate transcript-level counts `Salmon` will be used using the reference transcriptome as an input. Transcript level counts can be converted to gene-level counts using `scripts/salmon_tximport.r` which uses the `bioconductor::tximport` package. 
To perform normalization of the raw counts, differential gene expression analysis, and gene set enrichment analysis, use `scripts/dge_salmon.r` and `scripts/enrich_anal.svp.R` 

## Input
